### ISSUE TITLE
Possible race condition with gnome-gamma-tool + my brightness script on gnome

### ISSUE THREAD

**Issue by BinaryGenius** [zb3/gnome-gamma-tool#3]
[ggt-log.txt](https://github.com/zb3/gnome-gamma-tool/files/8819357/ggt-log.txt)
_I tried to organize this a bit but you can tell me if it's good or not._
# Preface
A lot of this file is redundant and I wasn't sure which information was useful vs which information was unnecessary so I opted to keep it all in this can be shortened later if it becomes an issue. It's a little above 17 thousand lines but it can be easily skimmed for some errors.
# Steps to reproduce
Map the command `sh /home/(username)/Projects/Gnome-Brightness-Ctl/brightness.sh Down` to F9 in gnome keyboard settings.
Map the command `sh /home/(username)/Projects/Gnome-Brightness-Ctl/brightness.sh Up` to F10 in gnome keyboard settings.
Recreate the directory structure and paste my brightness script in that directory along with ggt(gnome-gamma-tool).
Spam the F9 and F10 keys or hold them down and the screen starts to freak out.
# Theory
This is probably due to a race condition as mentioned here https://github.com/zb3/gnome-gamma-tool/discussions/2#discussioncomment-2749057
# Solutions
Find a way to limit the amount of times a user can run the script via the keybinds in a given time frame. However this is not ideal since the user should be able to change their brightness at will and as fast as possible.

**Comment by zb3** [comment]
After taking a look, I can say that there are at least three race conditions here.

### Race condition in gnome-gamma-tool
The first one is obvious - only one instance of gnome-gamma-tool can be running at a time. I fixed this in https://github.com/zb3/gnome-gamma-tool/commit/80c5ef7ba57f08bbe23ccc9dccaa7c2c842f1bf7
Note that the "speed" of this tool is limited by colord itself, so there's nothing I could do to make it faster here..

### Race condition in the script
The second problem is with your script - the same limitation applies because file operations are not atomic, so two instances of your script cannot be running at the same time. Something like this at the begining of the script should work, but I didn't test it:
```sh
exec 200 > "/tmp/.brightness.lock"
flock -n 200 || exit 1
```

### Unidentified race condition in colord(?)
Unfortunately even with these fixes, if you press these keys "fast enough" (whatever that means), sometimes this will lock for a few seconds, and you will see this error:
```
    new_profile = self.cd.import_profile_sync(Gio.File.new_for_path(tmp_path))
gi.repository.GLib.GError: cd_client_error: The profile was not added in time (0)
```

The bug is not within gnome-gamma-tool and there are multiple reported bugs, open for years but still unsolved:
https://bugzilla.redhat.com/show_bug.cgi?id=1372046
https://bugs.freedesktop.org/show_bug.cgi?id=82770
https://github.com/hughsie/colord/issues/94
https://bugs.launchpad.net/ubuntu/+source/unity-control-center/+bug/1366381

I tried running `gsd-color --verbose` and I saw this:
```
(gsd-color:37476): libcolord-DEBUG: 15:10:35.622: Incorrect content type for /home/zb3/.local/share/icc/gnome-gamma-tool-21d353ae-57a4-425b-9877-f6c8f4a3a113, got text/plain
```

But the problem was not with the generated file, since I was able to import the exact same file later (I even copied it just before calling import to make sure it's the same one). 

Unfortunately with cases like these, someone needs to dedicate weeks of work to dig through all of this, this time I certainly can't do that.

But if you have any workaround to make gnome-gamma-tool deal with this, let me know. With many GNOME bugs, workarounds are the only solutions :)

**Comment by BinaryGenius** [comment]
SO I had to learn what atomic meant, and what the heck were lock files but they seem like exactly what I needed. It locks the process for a set time so it can't be executed a hundred million times over itself. The brightness is not as fast but it is definitely way more stable and I haven't had it break on me after adding what you said.
```
The one issue was...
exec 200 > "/tmp/.brightness.lock"
flock -n 200 || exit 1

So I changed it to...
exec 200>"/tmp/.brightness.lock"
flock 200 || exit
```
I don't know what the "1" proceeding "exit" and the "-n" did so I removed them and the script worked fine. Also I wasn't talking about making the code blazing fast I just want it to be optimized (best possible code). Like if there is a fix to make things faster and me or you know it then add it to the code.

> ### Unidentified race condition in colord(?)
> 
> Unfortunately even with these fixes, if you press these keys "fast enough" (whatever that means), sometimes this will lock for a few seconds, and you will see this error:
> 
> ```
>     new_profile = self.cd.import_profile_sync(Gio.File.new_for_path(tmp_path))
> gi.repository.GLib.GError: cd_client_error: The profile was not added in time (0)
> ```
> 
> The bug is not within gnome-gamma-tool and there are multiple reported bugs, open for years but still unsolved: https://bugzilla.redhat.com/show_bug.cgi?id=1372046 https://bugs.freedesktop.org/show_bug.cgi?id=82770 [hughsie/colord#94](https://github.com/hughsie/colord/issues/94) https://bugs.launchpad.net/ubuntu/+source/unity-control-center/+bug/1366381
> 
> I tried running `gsd-color --verbose` and I saw this:
> 
> ```
> (gsd-color:37476): libcolord-DEBUG: 15:10:35.622: Incorrect content type for /home/zb3/.local/share/icc/gnome-gamma-tool-21d353ae-57a4-425b-9877-f6c8f4a3a113, got text/plain
> ```

I think the previous log showed that error but with your fixes the script does not show that error whatsoever it works fairly well. The file is the same as before but it has the fixes I stated plus the logging we talked about prior.

# Files
[ggt-log.txt](https://github.com/zb3/gnome-gamma-tool/files/8837353/ggt-log.txt)
[brightness.sh.txt](https://github.com/zb3/gnome-gamma-tool/files/8837359/brightness.sh.txt)

**Comment by zb3** [comment]
Glad it worked, in my case this error appears only when I press those keys "furiously" :)
Also yes I definitely made a typo with that "exec 200" line, sorry about that.

The `-n` argument is about so called "non-blocking" lock - when the lock is already set, the flock operation will return an error. When the `-n` argument is not present, the flock statement will instead wait till the lock is released. When implementing this inside gnome-gamma-tool, I used non-blocking locks just to avoid the case where many operations are already scheduled, like when the user presses the key many times because nothing happens, then the gamma is set continuously and the user can't cancel this process and set the new value. With non-blocking locks, if the program is already running, no operation is queued.

You can see the effect of this parameter by using this example script:

with `-n`:
```bash
#!/bin/sh

exec 200>"/tmp/.brightness.lock"
flock -n 200 || { echo already running && exit; }
echo now run this in another terminal

sleep 20
```

without `-n`:
```bash
#!/bin/sh

exec 200>"/tmp/.brightness.lock"
flock 200 || { echo already running && exit; }
echo now run this in another terminal

sleep 20
```

Let me know if I should close the issue.

**Comment by BinaryGenius** [comment]
Honestly that still makes zero sense I don't see any difference between both options nor does the log have any differences.

**Comment by zb3** [comment]
Oh, I'm not very good at explaining things, but let me try one more time with a video explanation :D
Note that for your script, you most likely won't observe any practical difference since most cases are handled by gnome-gamma-tool already.

So, this is the blocking version, without the `-n` parameter:

https://user-images.githubusercontent.com/3158267/172046018-0ddefce5-8b88-4d75-9a3e-3878872cd972.mp4


And this is the nonblocking one, with the `-n` parameter:

https://user-images.githubusercontent.com/3158267/172046029-4525c11b-27b5-452f-a4d8-b1e8cee69737.mp4

**Comment by BinaryGenius** [comment]
I tried using inspect element and downloading the video but to no avail it appears that your files are corrupted. Or weren't uploaded correctly
Image 1
![Screenshot from 2022-06-05 15-26-43](https://user-images.githubusercontent.com/105467963/172067297-74b0fe0a-fe3b-43f4-abfd-e58b2e97cecb.png)
Image 2
![Screenshot from 2022-06-05 15-27-16](https://user-images.githubusercontent.com/105467963/172067288-19bdc56c-d94e-4a2d-bca1-4f7b5b1ea113.png)

**Comment by zb3** [comment]
If you're using chromium in Fedora, you probably need the "chromium-freeworld" version from rpmfusion, because the H.264 standard is proprietary.. too bad GitHub didn't let me upload (much smaller) webm files :(

It works on mobile though, even on Firefox on LineageOS, so it should work on Android too

**Comment by BinaryGenius** [comment]
I use hardened firefox and I tried on tor still doesn't work you could explain it in text if it comes to that or just reupload the files via other means if github doesn't work. Look I'll try this on brave to give you the benefit of the doubt but I don't think it will work. It worked on brave luckily they had it on flathub, so I just downloaded and pasted the link for this issue then downloaded the videos.

So basically with the -n argument when you run the script while flock is active it just says "no" however with the -n it says error but since you had "||" the script was like "I'll run the first command or the second command but only if flock says it had an error" without the -n flock doesn't communicate at all it just stops execution of the script.

**Comment by zb3** [comment]
Yep, and without the `-n`, the execution is stopped until the other process exits, but the `flock` call doesn't fail so the part after `||` is not executed.

Good part about flocks is that you don't need to unlock them manually as these locks are automatically removed when the process exits. Back when I was using Arch Linux, its package manager `pacman` used a scheme where just the existence of the lock file meant the lock is taken. The result was that each time pacman crashed (very often), I had to remove the file manually or it wouldn't run. This won't happen with flock.

**Comment by BinaryGenius** [comment]
Alright well would you like to do a pull request or something like that I think that's what people do if they want to merge or add code. Not saying I'm a master coder or anything but I think it would add nicely to your code. You can close the issue after responding the issue's been solved and I'm clear on what flock -n does.

**Comment by zb3** [comment]
I'm not sure what you're referring to... the race condition was fixed in gnome-gamma-tool by this commit: https://github.com/zb3/gnome-gamma-tool/commit/80c5ef7ba57f08bbe23ccc9dccaa7c2c842f1bf7
That commit is already merged.

It's just that even if the race condition no longer occurs in gnome-gamma-tool, it might still occur in other scripts that use it, hence I suggested you could use `flock` in your script as a possible remedy. But that was actually out of scope, because it's not a gnome-gamma-tool issue.

**Comment by zb3** [comment]
I acknowledge that your use-case exposes a limitation of gnome-gamma-tool - you can't use ggt for keybindings because it doesn't support relative state changes. ggt should support that in the future (I have no energy to implement that now).

But the script is a separate tool that builds **on top** of gnome-gamma-tool. Hence you should create your own repository with your script (I see no problem with you putting a copy of ggt there) so others can use it.

**Comment by BinaryGenius** [comment]
Yeah you're right my script is a bit botched together with ggt, I wont worry about it for now but what I will do is open a discussion for something that I think is more pressing which is the the migration of this script from gnome 41 to 42.
